test/redis_test.rb

```diff
@@ -0,0 +1,180 @@
+  end
+
+  def test_connect_instrumentation
+    notified = false
+    subscriber = Semian.subscribe do |event, resource, scope|
+      notified = true
+      assert_equal :success, event
+      assert_equal Semian[:redis_testing], resource
+      assert_equal :connect, scope
+    end
+
+    connect_to_redis!
+
+    assert notified, 'No notification have been emitted'
+  ensure
+    Semian.unsubscribe(subscriber)
+  end
+
+  def test_resource_acquisition_for_connect
+    client = connect_to_redis!
+
+    Semian[:redis_testing].acquire do
+      assert_raises Redis::ResourceOccupiedError do
+        connect_to_redis!
+      end
+    end
+  end
+
+  def test_resource_timeout_on_connect
+    @proxy.downstream(:latency, latency: 500).apply do
+      background { connect_to_redis! }
+
+      assert_raises Redis::ResourceOccupiedError do
+        connect_to_redis!
+      end
+    end
+  end
+
+  def test_circuit_breaker_on_connect
+    @proxy.downstream(:latency, latency: 500).apply do
+      background { connect_to_redis! }
+
+      ERROR_THRESHOLD.times do
+        assert_raises Redis::ResourceOccupiedError do
+          connect_to_redis!
+        end
+      end
+    end
+
+    yield_to_background
+
+    assert_raises Redis::CircuitOpenError do
+      connect_to_redis!
+    end
+
+    Timecop.travel(ERROR_TIMEOUT + 1) do
+      connect_to_redis!
+    end
+  end
+
+  def test_query_instrumentation
+    client = connect_to_redis!
+
+    notified = false
+    subscriber = Semian.subscribe do |event, resource, scope|
+      notified = true
+      assert_equal :success, event
+      assert_equal Semian[:redis_testing], resource
+      assert_equal :io, scope
+    end
+
+    client.get('foo')
+
+    assert notified, 'No notification have been emitted'
+  ensure
+    Semian.unsubscribe(subscriber)
+  end
+
+  def test_resource_acquisition_for_query
+    client = connect_to_redis!
+
+    Semian[:redis_testing].acquire do
+      assert_raises Redis::ResourceOccupiedError do
+        client.get('foo')
+      end
+    end
+  end
+
+  def test_resource_timeout_on_query
+    client = connect_to_redis!
+    client2 = connect_to_redis!
+
+    @proxy.downstream(:latency, latency: 500).apply do
+      background { client2.get('foo') }
+
+      assert_raises Redis::ResourceOccupiedError do
+        client.get('foo')
+      end
+    end
+  end
+
+  def test_circuit_breaker_on_query
+    client = connect_to_redis!
+    client2 = connect_to_redis!
+
+    client.set('foo', 2)
+
+    @proxy.downstream(:latency, latency: 1000).apply do
+      background { client2.get('foo') }
+
+      ERROR_THRESHOLD.times do
+        assert_raises Redis::ResourceOccupiedError do
+          client.get('foo')
+        end
+      end
+    end
+
+    yield_to_background
+
+    assert_raises Redis::CircuitOpenError do
+      client.get('foo')
+    end
+
+    Timecop.travel(ERROR_TIMEOUT + 1) do
+      assert_equal '2', client.get('foo')
+    end
+  end
+
+  private
```

Comment:
Needs a test on the nested acquire (although I'm still not sure what that'll look like).